extension/src/format/parser.ts

```diff
@@ -182,4 +182,149 @@ export class LispParser {
             this.atomsForest.push(lispLists);
         }
     }
+}
+
+
+export namespace LispParser {
+    
+    enum ParseState {
+        UNKNOWN = 0,
+        STRING = 1,
+        COMMENT = 2
+    }
+
+    interface CountTracker {
+        idx: number;
+        line: number;
+        column: number;
+        linefeed: string;
+    }
+
+    export function getDocumentSexpression(data: string|vscode.TextDocument, tracker?: CountTracker): Sexpression {
+        let documentText = '';
+        if (typeof(data) === 'string') {
+            documentText = data;
+        } else {
+            documentText = data.getText();
+        }
+        if (!tracker){
+            tracker = {
+                idx: 0, line: 0, column: 0, 
+                linefeed: typeof(data) === 'string' ? (data.indexOf('\r\n') >= 0 ? '\r\n' : '\n') : LispParser.getEOL(data)
+            };
+        }
+        
+        const result = new Sexpression();
+        result.line = tracker.line;
+        result.column = tracker.column;
+        result.linefeed = tracker.linefeed;
+        
+        let isAuthorized = true;
+        let state = ParseState.UNKNOWN;
+        let grpStart: vscode.Position = null;
+        let temp = '';
+        
+        while (tracker.idx < documentText.length && isAuthorized) {
```

Comment:
I think this is duplicate with the current LispParser, and you plan to replace the current LispParser with this new one? We should keep the codes consistency, to follow the [DRY](https://dzone.com/articles/software-design-principles-dry-and-kiss) rule.

Reply:
If you want me to do all this in the shadows and only bring it out after I can replace the whole thing, then I will..... but I thought this was too fundamental to operate in that way.

Reply:
It looks like the old lisp parser has 10x slow down performance issue than this one you written. I will take some time to do profiling. 
If we decided to replace the old lisp parser, it needs to do lots of regression tests for format (document format, selection format and the enter key format, and the project file format).
If you also think that it is large task, I agree that we hold on this PR, and focus on other tasks. (And this is fundamental, I think that it is worthy to take more time to refine/improve it).

Reply:
I am of the opinion that (yes) this is very similar, but (no) it provides entirely different use cases from the `LispParser.AtomsForest`

I say that because of an attempt to use this because the `LispFormatter.formatGut()` does appear to be dependent on raw string values that the AtomsForest provides and I have been saying for several PR's I need something with LispAtom's not strings.